// @storybook/vue3/dist/cjs/client/preview/render.js
"use strict";

require("core-js/modules/es.array.slice.js");

require("core-js/modules/es.object.freeze.js");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = render;
exports.storybookApp = void 0;

require("core-js/modules/es.function.name.js");

require("core-js/modules/es.array.concat.js");

var _tsDedent = _interopRequireDefault(require("ts-dedent"));

var _vue = require("vue");

var _templateObject;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

var activeStoryComponent = (0, _vue.shallowRef)(null);
var root = null;
var storybookApp = (0, _vue.createApp)({
  // If an end-user calls `unmount` on the app, we need to clear our root variable
  unmounted: function unmounted() {
    root = null;
  },
  setup: function setup() {
    return function () {
      if (!activeStoryComponent.value) throw new Error('No Vue 3 Story available. Was it set correctly?');
      return (0, _vue.h)(activeStoryComponent.value);
    };
  }
});
exports.storybookApp = storybookApp;

function render(_ref) {
  var storyFn = _ref.storyFn,
      kind = _ref.kind,
      name = _ref.name,
      args = _ref.args,
      showMain = _ref.showMain,
      showError = _ref.showError,
      showException = _ref.showException,
      forceRender = _ref.forceRender;
  storybookApp.config.errorHandler = showException;
  var element = storyFn();

  if (!element) {
    showError({
      title: "Expecting a Vue component from the story: \"".concat(name, "\" of \"").concat(kind, "\"."),
      description: (0, _tsDedent.default)(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n        Did you forget to return the Vue component from the story?\n        Use \"() => ({ template: '<my-comp></my-comp>' })\" or \"() => ({ components: MyComp, template: '<my-comp></my-comp>' })\" when defining the story.\n      "])))
    });
    return;
  }

  showMain();
  activeStoryComponent.value = element;

  if (!root) {
    root = storybookApp.mount('#root');
  }
}
